pkg/modules/notary/notary.go

```diff
@@ -0,0 +1,300 @@
+		if _, ok := n.requests[tx.Hash()]; ok {
+			delete(n.requests, tx.Hash()) // main transaction is accepted, so request can be removed
+		}
+	}
+	for h, r := range n.requests {
+		if !r.isSent && r.isValid && r.nKeys == r.nKeysCollected && r.minNotValidBefore > bc.BlockHeight() {
+			if err := n.finalize(r.main); err != nil {
+				n.Config.Log.Error("failed to finalize main transaction", zap.Error(err))
+			} else {
+				r.isSent = true
+			}
+			continue
+		}
+		if r.minNotValidBefore <= bc.BlockHeight() { // then at least one of the fallbacks can already be sent.
+			newFallbacks := r.fallbacks[:0]
+			for _, fb := range r.fallbacks {
+				if nvb := fb.GetAttributes(transaction.NotValidBeforeT)[0].Value.(*transaction.NotValidBefore).Height; nvb <= bc.BlockHeight() {
+					if err := n.finalize(fb); err != nil {
+						newFallbacks = append(newFallbacks, fb) // wait for the next block to resend them
+					}
+				} else {
+					newFallbacks = append(newFallbacks, fb)
+				}
+			}
+			if len(newFallbacks) == 0 {
+				delete(n.requests, h)
+			} else {
+				r.fallbacks = newFallbacks
+			}
+		}
+	}
+}
+
+// finalize adds missing Notary witnesses to the transaction (main or fallback) and pushes it to the network.
+func (n *Notary) finalize(tx *transaction.Transaction) error {
+	acc := n.getAccount()
+	if acc == nil {
+		return errors.New("no available Notary account")
+	}
+	w := io.NewBufBinWriter()
+	emit.Bytes(w.BinWriter, acc.PrivateKey().PublicKey().Bytes())
+	notaryWitness := transaction.Witness{
+		InvocationScript:   w.Bytes(),
+		VerificationScript: []byte{},
+	}
+	for i, signer := range tx.Signers {
+		if signer.Account == n.Config.Chain.GetNotaryContractScriptHash() {
+			tx.Scripts[i] = notaryWitness
+			break
+		}
+	}
+	return n.getOnTransaction()(tx)
+}
+
+// SetOnTransaction sets callback method which should be called after transaction is completed.
+func (n *Notary) SetOnTransaction(f func(tx *transaction.Transaction) error) {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+	n.onTransaction = f
+}
+
+func (n *Notary) getOnTransaction() func(tx *transaction.Transaction) error {
+	n.mtx.RLock()
+	defer n.mtx.RUnlock()
+	return n.onTransaction
+}
+
+// verifyIncompleteWitnesses checks that tx either doesn't have all witnesses attached (in this case none of them
+// can be multisignature), or it only has a partial multisignature.
+func (n *Notary) verifyIncompleteWitnesses(t *transaction.Transaction, nKeys uint8) (RequestType, error) {
+	var (
+		typ         RequestType
+		nKeysActual uint8
+	)
+	for i, w := range t.Scripts {
+		if t.Signers[i].Account == n.Config.Chain.GetNotaryContractScriptHash() || len(w.InvocationScript) == 0 { // allow empty witness for notary contract or empty dummy witness
+			continue
+		}
+		if typ == Signature || typ == MultiSignature {
+			return 0, fmt.Errorf("bad type of witness #%d: only one partial witness is allowed", i)
```

Comment:
You can have multiple `Signature`s missing, that's the point.

Reply:
Fixed.